### PR TITLE
Plugins: Fix sorting issue with expandable rows

### PR DESCRIPTION
**Pull request by fabrizio-grafana**
When rows are expandable and sortable, the two features currently clash, as reported in [this escalation](https://github.com/orgs/grafana/projects/221/views/1?filterQuery=assignee%3A%40me+-status%3Adone&pane=issue&itemId=37147025). This PR fixes such issue.

If we like this approach, we can then iterate to add tests.

Fixes https://github.com/grafana/grafana/issues/74443

### CODE REVIEW

**Comment by oscarkilhed** [comment]
Would be nice to have a test for this since manual testing with nested data is a bit of a hassle currently.  `Table.test.tsx` has a test for nested data that would be fairly good as a starting point.

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 1]
```diff
@@ -110,8 +110,25 @@ export const Table = memo((props: Props) => {
     [data, width, columnMinWidth, footerItems, hasNestedData, isCountRowsSet]
   );
 
+  const toggleAllRowsExpandedRef = useRef<(value?: boolean) => void>();
```
Comment (aocenas) [inline packages/grafana-ui/src/components/Table/Table.tsx:113]: I think this needs a bit more explanation either in code comments or in the PR of why do we need the ref for the function here as it's not right away clear.

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 2]
```diff
@@ -142,9 +159,16 @@ export const Table = memo((props: Props) => {
...
     pageOptions,
+    toggleAllRowsExpanded,
   } = useTable(options, useFilters, useSortBy, useAbsoluteLayout, useResizeColumns, useExpanded, usePagination);
 
   const extendedState = state as GrafanaTableState;
+  toggleAllRowsExpandedRef.current = toggleAllRowsExpanded;
+
+  const expandedRowsRepr = JSON.stringify(Object.keys(state.expanded));
+  useEffect(() => {
+    listRef.current?.resetAfterIndex(0);
```
Comment (aocenas) [inline packages/grafana-ui/src/components/Table/Table.tsx:170]: I assume that this means, every time we change what is expanded, we recompute the variable sized list so the heights are correct?

File: packages/grafana-ui/src/components/Table/Table.tsx [hunk 3]
```diff
@@ -110,8 +110,25 @@ export const Table = memo((props: Props) => {
...
-  const stateReducer = useTableStateReducer(props);
+  const stateReducer = useTableStateReducer({
+    ...props,
+    onSortByChange: (state) => {
+      // Collapse all rows. This prevents a known bug that causes the size of the rows to be incorrect due to
+      // using `VariableSizeList` and `useExpanded` together.
+      if (toggleAllRowsExpandedRef.current) {
+        toggleAllRowsExpandedRef.current(false);
+      } else {
+        console.error('toggleAllRowsExpandedRef.current is undefined');
```
Comment (aocenas) [inline packages/grafana-ui/src/components/Table/Table.tsx:124]: This is a bit weird. Like if this happens, is it a programming error? If not can we recover? It's not clear from this if this is something that user/we can ignore or something to fix in the future.

**Review by aocenas — APPROVED** [review]

**Pull request merged**